src/probability/upcrossing.lean

```diff
@@ -0,0 +1,897 @@
+/-
+Copyright (c) 2022 Kexing Ying. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kexing Ying
+-/
+import probability.hitting_time
+import probability.martingale
+
+/-!
+
+# Doob's upcrossing estimate
+
+Given a discrete real-valued submartingale $(f_n)_{n \in \mathbb{N}}$, denoting $U_N(a, b)$ the
+number of times $f_n$ crossed from below $a$ to above $b$ before time $N$, Doob's upcrossing
+estimate (also known as Doob's inequality) states that
+$$(b - a) \mathbb{E}[U_N(a, b)] \le \mathbb{E}[(f_N - a)^+].$$
+Doob's upcrossing estimate is an important inequality and is central in proving the martingale
+convergence theorems.
+
+## Main definitions
+
+* `measure_theory.upper_crossing a b f N n`: is the stopping time corresponding to `f` crossing
+  above `b` the `n`-th time.
+* `measure_theory.lower_crossing a b f N n`: is the stopping time corresponding to `f` crossing
+  below `a` the `n`-th time.
+* `measure_theory.upcrossing_strat a b f N`: is the predicatable process which is 1 if `n` is
+  between a consecutive pair of lower and upper crossing and is 0 otherwise. Intuitively
+  one might think of the `upcrossing_strat` as the strategy of buying 1 share whenever the process
+  crosses below `a` for the first time after selling and selling 1 share whenever the process
+  crosses above `b` for the first time after buying.
+* `measure_theory.upcrossing_before a b f N`: is the number of times `f` crosses from below `a` to
```

Comment:
`upcrossings_before`?

Reply:
I changed the name of `upcrossing` to `upcrossings` also